src/maxflow.rs

```diff
@@ -1 +1,325 @@
+            if d <= Cap::zero() {
+                continue;
+            }
+            self.graph.g[v][i].cap += d;
+            self.graph.g[e_to][e_rev].cap -= d;
+            res += d;
+            if res == up {
+                break;
+            }
+        }
+        self.iter[v] = self.graph.g[v].len();
+        res
+    }
+}
+
+#[derive(Default)]
+pub struct MfGraph<Cap> {
+    _n: usize,
+    pos: Vec<(usize, usize)>,
+    g: Vec<Vec<_Edge<Cap>>>,
+}
+
+struct _Edge<Cap> {
+    to: usize,
+    rev: usize,
+    cap: Cap,
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{Edge, MfGraph};
+
+    #[test]
+    fn test_max_flow_wikipedia() {
+        // From https://commons.wikimedia.org/wiki/File:Min_cut.png
+        // Under CC BY-SA 3.0 https://creativecommons.org/licenses/by-sa/3.0/deed.en
+        let mut graph = MfGraph::new(6);
+        assert_eq!(graph.add_edge(0, 1, 3), 0);
+        assert_eq!(graph.add_edge(0, 2, 3), 1);
+        assert_eq!(graph.add_edge(1, 2, 2), 2);
+        assert_eq!(graph.add_edge(1, 3, 3), 3);
+        assert_eq!(graph.add_edge(2, 4, 2), 4);
+        assert_eq!(graph.add_edge(3, 4, 4), 5);
+        assert_eq!(graph.add_edge(3, 5, 2), 6);
+        assert_eq!(graph.add_edge(4, 5, 3), 7);
+
+        assert_eq!(graph.flow(0, 5), 5);
+
+        let edges = graph.edges();
+        {
+            #[rustfmt::skip]
+            assert_eq!(
+                edges,
+                vec![
+                    Edge { from: 0, to: 1, cap: 3, flow: 3 },
+                    Edge { from: 0, to: 2, cap: 3, flow: 2 },
+                    Edge { from: 1, to: 2, cap: 2, flow: 0 },
+                    Edge { from: 1, to: 3, cap: 3, flow: 3 },
+                    Edge { from: 2, to: 4, cap: 2, flow: 2 },
+                    Edge { from: 3, to: 4, cap: 4, flow: 1 },
+                    Edge { from: 3, to: 5, cap: 2, flow: 2 },
+                    Edge { from: 4, to: 5, cap: 3, flow: 3 },
+                ]
+            );
+        }
+        assert_eq!(
+            graph.min_cut(0),
+            vec![true, false, true, false, false, false]
+        );
+    }
+
+    #[test]
+    fn test_max_flow_wikipedia_multiple_edges() {
+        // From https://commons.wikimedia.org/wiki/File:Min_cut.png
+        // Under CC BY-SA 3.0 https://creativecommons.org/licenses/by-sa/3.0/deed.en
+        let mut graph = MfGraph::new(6);
+        for &(u, v, c) in &[
+            (0, 1, 3),
+            (0, 2, 3),
+            (1, 2, 2),
+            (1, 3, 3),
+            (2, 4, 2),
+            (3, 4, 4),
+            (3, 5, 2),
+            (4, 5, 3),
+        ] {
+            for _ in 0..c {
+                graph.add_edge(u, v, 1);
+            }
+        }
+
+        assert_eq!(graph.flow(0, 5), 5);
+        assert_eq!(
+            graph.min_cut(0),
+            vec![true, false, true, false, false, false]
+        );
+    }
+
+    #[test]
+    #[allow(clippy::many_single_char_names)]
+    fn test_max_flow_misawa() {
```

Comment:
You might have done already, but verified locally that it gonna take very long for n=20 and take forever when we deleted L:189 and L:209 as intended :)